interface/java_binding/src/main/cpp/KTXTexture.cpp

```diff
@@ -0,0 +1,287 @@
+/*
+ * Copyright (c) 2021, Shukant Pal and Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <assert.h>
+#include <cstring>
+#include <jni.h>
+#include <vector>
+#include <iostream>
+
+#include "ktx.h"
+#include "libktx-jni.h"
+
+struct pinned_image_buf
+{
+    jbyteArray handle;
+    jbyte *data;
+};
+
+// The buffer list is a vector of memory buffers pinned by KTXTexture. These buffers are pinned
+// when setImageFromMemory is used. They are freed when KTXTexture.destroy is invoked.
+
+static std::vector<pinned_image_buf> *get_or_create_buffer_list(JNIEnv *env, jobject thiz)
+{
+    jclass ktx_texture_class = env->GetObjectClass(thiz);
+    jfieldID ktx_buffers_field = env->GetFieldID(ktx_texture_class, "buffers", "J");
+
+    std::vector<pinned_image_buf> *buffers =
+        reinterpret_cast<std::vector<pinned_image_buf>*>(env->GetLongField(thiz, ktx_buffers_field));
+
+    // Lazy init
+    if (buffers == NULL) {
+        buffers = new std::vector<pinned_image_buf>();
+        env->SetLongField(thiz, ktx_buffers_field, reinterpret_cast<jlong>(buffers));
+    }
+
+    return buffers;
+}
+
+static void push_buffer_list(JNIEnv *env, jobject thiz, jbyteArray handle, jbyte *data)
+{
+    std::vector<pinned_image_buf> *buffers = get_or_create_buffer_list(env, thiz);
+
+    pinned_image_buf buf;
+
+    buf.handle = handle;
+    buf.data = data;
+
+    buffers->push_back(buf);
+}
+
+static void free_buffer_list(JNIEnv *env, jobject thiz)
+{
+    jclass ktx_texture_class = env->GetObjectClass(thiz);
+    jfieldID ktx_buffers_field = env->GetFieldID(ktx_texture_class, "buffers", "J");
+
+    std::vector<pinned_image_buf> *buffers =
+        reinterpret_cast<std::vector<pinned_image_buf>*>(env->GetLongField(thiz, ktx_buffers_field));
+
+    // Nothing to free
+    if (buffers == NULL) {
+        return;
+    }
+
+    std::vector<pinned_image_buf> l_buffers = *buffers;
+
+    for (std::vector<pinned_image_buf>::iterator it = std::begin(l_buffers);
+        it != std::end(l_buffers);
+        ++it)
+    {
+        pinned_image_buf buffer = *it;
+
+        env->ReleaseByteArrayElements(buffer.handle, buffer.data, JNI_ABORT);
+    }
+
+    env->SetLongField(thiz, ktx_buffers_field, 0);
+    delete buffers;
+}
+
+extern "C" JNIEXPORT jint JNICALL Java_org_khronos_ktx_KTXTexture_getBufferListSize(JNIEnv *env, jobject thiz)
```

Comment:
In Java code, classes aren't referred to by the canonical name (generally); instead, the classes are imported:

```java
import org.khronos.ktx.KTXTexture;

public class Main {
  public static void main(String[] args) {
     KTXTexture tex; // not org.khronos.ktx.KTXTexture
  }
}
```

If another library (say libGDX) already has a `Texture` class, developers would have to refer to libktx's texture with `org.khronos.ktx.Texture`, which is painful. Since all of the classes are KTX-specific anyway, I prefixed them.

> If so, why not match the C code prefix?

I though this did match the C code prefix 😅. How do you imagine I do it?